### ISSUE TITLE
[BUG] http json falling on java.lang.OutOfMemoryError: Java heap space.

### ISSUE THREAD

**Issue by ShlomoNus** [digital-asset/daml#12248]
# Affected Daml version

1.14 version.

# Bug description

we are using Daml about a year, and lately we have a bug with http json as mentioned in the title, we did not added any changes to ledger and it's traffic, so we find it hard to figure the reason for the bug, do you have any Idea why that may be happening ?

# Expected behavior

http json keep working as before.

# Additional context

We are using AWS cloud platform.

**Comment by stefanobaghino-da** [comment]
Are you using the database-backed query store or are you relying on the in-memory one?

**Comment by ShlomoNus** [comment]
yes. we are using PostgreSQL-based Daml ledger implementation. we are using the release 1.14.0 : daml-on-sql-1.14.0.jar, http-json-1.14.0.jar, trigger-service-1.14.0.jar

**Comment by stefanobaghino-da** [comment]
Please note that 1.14 is quite dated and I would recommend upgrading to a later version: there's a chance this is already fixed and if it's not it's unlikely we're going to back port the fix to 1.14.

If by any chance the issue persists with 1.18 (the latest stable version as of the time of writing), I have a few extra questions that could help us understand how this happens:
* How long after upgrading to 1.14 did you experience this?
* From what version did you upgrade?
* Is the system load equivalent to the one you had before the upgrade? 
* Are there specific triggering causes that we should look into?
* Does this happen after some more or less fixed amount of time or in response to a spike in load?
* How many active streams and/or ongoing queries do you have at the time of the OOM?

**Comment by ShlomoNus** [comment]
thanks for quick response Stefan, we will upgrade.
I found out from our team lead we are not using any db for caching queries, in memory only,  does it make any difference?Thank you in advance

**Comment by realvictorprm** [comment]
> thanks for quick response Stefan, we will upgrade. I found out from our team lead we are not using any db for caching queries, in memory only, does it make any difference?Thank you in advance

Hi @ShlomoNus, yeah there is a big difference between using the in-memory store vs the db store. I highly recommend using the db store to avoid running into the OOM error because the store will inevitably only grow over time with the db most likely having much more memory available :wink:

**Comment by stefanobaghino-da** [comment]
> thanks for quick response Stefan, we will upgrade. I found out from our team lead we are not using any db for caching queries, in memory only, does it make any difference?Thank you in advance

Thanks for getting back to us, I heavily suspect that using the in-memory implementation is the source of the issue, as mentioned bu @realvictorprm in the previous message. If you want to keep using that, one approach is to expect them to blow up in size and use an orchestration system like Kubernetes that manages live replicas (blunt, but works). This also means that you lose the cache over the restart and you'll need to recover state from the ledger, which depending on your application, could also mean that the application will simply keep crashing as the state needed by the application exceeds the limits of your JVM's memory.

The cleaner solution is to use a database-backed query store, which could also give you better performance thanks to the indexing performed by the database (but this is not necessarily guaranteed and heavily depends on your workload).

The query store setup and configuration is described on our documentations ([link](https://docs.daml.com/1.18.0/json-api/production-setup.html#query-store)).

I'll close the issue for now, but please do let us know if the issue still occurs using the database-backed query store.

**Comment by shiranA90** [comment]
Following your suggestions, we have upgraded to version 1.18.0 and configured a PostgreSQL backend as a cache but we are still having the same problem.
I have attached a few photos which show the error logs, metrics for Gauges, and the memory utilization and CPU utilization which keeps track of changes in http-json API after the changes we did.
Thank you in advance
![CPU_utillization](https://user-images.githubusercontent.com/66947619/148682577-cca92fad-cfc0-4504-b8cd-b848435c004c.jpg)
![memory_utillization](https://user-images.githubusercontent.com/66947619/148682583-a228f295-9168-4b87-b3a9-a33bc6b8945b.jpg)
![error_logs](https://user-images.githubusercontent.com/66947619/148682587-8ed9de3b-9099-4d03-816b-0c676a2d0e68.jpg)
![gauges_part1](https://user-images.githubusercontent.com/66947619/148682594-febebdc8-2215-45af-a434-bc754e9779ec.jpg)
![gauges_part2](https://user-images.githubusercontent.com/66947619/148682598-ee3b132b-fc28-4572-b58c-9b04a4029bd1.jpg)
![gauges_part3](https://user-images.githubusercontent.com/66947619/148682600-0da57b97-9934-49c4-895d-dbe34ef9a124.jpg)
![gauges_part4](https://user-images.githubusercontent.com/66947619/148682602-4817989b-86f0-4835-9bdf-718aff26d1b0.jpg)
.

**Comment by stefanobaghino-da** [comment]
Thanks for the detailed report, we'll see if there's something that clearly stands out and can be fixed quickly or prioritize this after the next release if it requires more work.

**Comment by stefanobaghino-da** [comment]
I just noticed that it looks like you're only giving ~0.5 GB to the process. Are you sure this is enough? Have you tried giving the HTTP JSON API process more memory?

**Comment by stefanobaghino-da** [comment]
(I said "I noticed" but I should really point out that @da-tanabe did, thanks 🙇🏻)

**Comment by umachc** [comment]
![image](https://user-images.githubusercontent.com/97630855/149213754-e35ae47a-f95f-4c86-9f29-1e920eb11098.png)

**Comment by benm-da** [comment]
@umachc @shiranA90 

Please see the following response from @stefanobaghino-da via Zendesk:

"In order for memory to be reclaimed a garbage collection event must be triggered. The JVM does not trigger garbage collection events all the time (or possibly ever) if the memory reserved to the process is sufficient. If you monitor the JVM process, the memory usage will simply expand until the maximum amount is reached.

If you measure the usage from within the JVM, you should see a completely normal "sawtooth" shaped memory usage pattern in which the memory fills up until the maximum is reached and then drops after a garbage collection event. If you want to put my claim to the test, just keep running your test for an extended period of time."
